File: Новая папка/src/test/example/MainTest.py
import pytest
from Main import MathFunctions


class TestMathFunctions:
    """Тесты для математических функций"""

    @pytest.fixture
    def app(self):
        """Фикстура для создания экземпляра класса"""
        return MathFunctions()

    # Группа: MATH_OPERATIONS
    class TestMathOperations:
        """Тесты математических операций"""

        # Data providers через параметризацию pytest
        @pytest.mark.parametrize("number,expected", [
            (2, True),
            (3, True),
            (17, True),
            (15, False),
            (1, False),
            (0, False)
        ])
        @pytest.mark.math
        @pytest.mark.fast
        def test_is_prime_with_data_provider(self, app, number, expected):
            """Тест is_prime с параметризацией"""
            # ACT - ДЕЙСТВИЕ
            actual_result = app.is_prime(number)

            # ASSERT - СРАВНЕНИЕ ОЖИДАЕМОГО И ФАКТИЧЕСКОГО РЕЗУЛЬТАТОВ
            assert actual_result == expected, \
                f"Число {number} должно {'быть' if expected else 'не быть'} простым"

        @pytest.mark.math
        @pytest.mark.fast
        def test_is_prime(self, app):
            """Тест проверки простых чисел"""
            # ARRANGE - ПОДГОТОВКА ТЕСТОВОГО ОКРУЖЕНИЯ
            prime_number = 17
            non_prime_number = 15

            # ACT - ДЕЙСТВИЕ
            actual_prime_result = app.is_prime(prime_number)
            actual_non_prime_result = app.is_prime(non_prime_number)

            # ASSERT - СРАВНЕНИЕ ОЖИДАЕМОГО И ФАКТИЧЕСКОГО РЕЗУЛЬТАТОВ
            assert actual_prime_result, "17 должно быть простым числом"
            assert not actual_non_prime_result, "15 не должно быть простым числом"

        @pytest.mark.parametrize("number,expected", [
            (0, 1),
            (1, 1),
            (5, 120),
            (7, 5040)
        ])
        @pytest.mark.math
        @pytest.mark.fast
        def test_factorial_with_data_provider(self, app, number, expected):
            """Тест factorial с параметризацией"""
            # ACT - ДЕЙСТВИЕ
            actual_result = app.factorial(number)

            # ASSERT - СРАВНЕНИЕ ОЖИДАЕМОГО И ФАКТИЧЕСКОГО РЕЗУЛЬТАТОВ
            assert actual_result == expected, \
                f"Факториал {number} должен быть равен {expected}"

        @pytest.mark.math
        @pytest.mark.fast
        def test_factorial(self, app):
            """Тест вычисления факториала"""
            # ARRANGE - ПОДГОТОВКА ТЕСТОВОГО ОКРУЖЕНИЯ
            input_number = 5
            expected_result = 120

            # ACT - ДЕЙСТВИЕ
            actual_result = app.factorial(input_number)

            # ASSERT - СРАВНЕНИЕ ОЖИДАЕМОГО И ФАКТИЧЕСКОГО РЕЗУЛЬТАТОВ
            assert actual_result == expected_result, "Факториал 5 должен быть равен 120"

        # ГРУППА: MATH_OPERATIONS + EXCEPTION_TESTS
        @pytest.mark.math
        @pytest.mark.exception
        def test_factorial_negative(self, app):
            """Тест factorial с отрицательным числом"""
            # ARRANGE - ПОДГОТОВКА ТЕСТОВОГО ОКРУЖЕНИЯ
            negative_number = -1

            # ACT & ASSERT - ДЕЙСТВИЕ И ПРОВЕРКА ИСКЛЮЧЕНИЯ
            with pytest.raises(ValueError, match="Число не может быть отрицательным"):
                app.factorial(negative_number)

        @pytest.mark.math
        @pytest.mark.fast
        def test_fibonacci(self, app):
            """Тест чисел Фибоначчи"""
            # ARRANGE - ПОДГОТОВКА ТЕСТОВОГО ОКРУЖЕНИЯ
            inputs = [0, 1, 6, 10]
            expected_results = [0, 1, 8, 55]

            # ACT & ASSERT - ДЕЙСТВИЕ И ПРОВЕРКА ДЛЯ НЕСКОЛЬКИХ ЗНАЧЕНИЙ
            for i in range(len(inputs)):
                # ACT - ДЕЙСТВИЕ
                actual_result = app.fibonacci(inputs[i])

                # ASSERT - СРАВНЕНИЕ ОЖИДАЕМОГО И ФАКТИЧЕСКОГО РЕЗУЛЬТАТОВ
                assert actual_result == expected_results[i], \
                    f"Число Фибоначчи F({inputs[i]}) должно быть равно {expected_results[i]}"

        # ГРУППА: MATH_OPERATIONS + EXCEPTION_TESTS
        @pytest.mark.math
        @pytest.mark.exception
        def test_fibonacci_negative(self, app):
            """Тест fibonacci с отрицательным индексом"""
            # ARRANGE - ПОДГОТОВКА ТЕСТОВОГО ОКРУЖЕНИЯ
            negative_index = -5

            # ACT & ASSERT - ДЕЙСТВИЕ И ПРОВЕРКА ИСКЛЮЧЕНИЯ
            with pytest.raises(ValueError, match="Индекс не может быть отрицательным"):
                app.fibonacci(negative_index)

    # ГРУППА: ARRAY_OPERATIONS
    class TestArrayOperations:
        """Тесты операций с массивами"""

        @pytest.mark.array
        @pytest.mark.fast
        def test_find_max(self, app):
            """Тест поиска максимального числа"""
            # ARRANGE - ПОДГОТОВКА ТЕСТОВОГО ОКРУЖЕНИЯ
            numbers = [3, 7, 2, 9, 1]
            expected_result = 9

            # ACT - ДЕЙСТВИЕ
            actual_result = app.find_max(numbers)

            # ASSERT - СРАВНЕНИЕ ОЖИДАЕМОГО И ФАКТИЧЕСКОГО РЕЗУЛЬТАТОВ
            assert actual_result == expected_result, "Максимальное число должно быть 9"

        @pytest.mark.array
        @pytest.mark.fast
        def test_find_max_with_single_element(self, app):
            """Тест поиска максимального в массиве из одного элемента"""
            # ARRANGE - ПОДГОТОВКА ТЕСТОВОГО ОКРУЖЕНИЯ
            single_element_array = [5]
            expected_result = 5

            # ACT - ДЕЙСТВИЕ
            actual_result = app.find_max(single_element_array)

            # ASSERT - СРАВНЕНИЕ ОЖИДАЕМОГО И ФАКТИЧЕСКОГО РЕЗУЛЬТАТОВ
            assert actual_result == expected_result, \
                "Максимальное число в массиве из одного элемента должно быть этим элементом"

        # ГРУППА: ARRAY_OPERATIONS + EXCEPTION_TESTS
        @pytest.mark.array
        @pytest.mark.exception
        def test_find_max_empty_array(self, app):
            """Тест поиска максимального в пустом массиве"""
            # ARRANGE - ПОДГОТОВКА ТЕСТОВОГО ОКРУЖЕНИЯ
            empty_array = []

            # ACT & ASSERT - ДЕЙСТВИЕ И ПРОВЕРКА ИСКЛЮЧЕНИЯ
            with pytest.raises(ValueError, match="Массив не может быть пустым"):
                app.find_max(empty_array)

    # ГРУППА: STRING_OPERATIONS
    class TestStringOperations:
        """Тесты строковых операций"""

        @pytest.mark.parametrize("input_text,expected", [
            ("А роза упала на лапу Азора", True),
            ("racecar", True),
            ("hello", False),
            ("", True),
            ("a", True)
        ])
        @pytest.mark.string
        @pytest.mark.fast
        def test_is_palindrome_with_data_provider(self, app, input_text, expected):
            """Тест is_palindrome с параметризацией"""
            # ACT - ДЕЙСТВИЕ
            actual_result = app.is_palindrome(input_text)

            # ASSERT - СРАВНЕНИЕ ОЖИДАЕМОГО И ФАКТИЧЕСКОГО РЕЗУЛЬТАТОВ
            assert actual_result == expected, \
                f"Строка '{input_text}' должна {'быть' if expected else 'не быть'} палиндромом"

        @pytest.mark.string
        @pytest.mark.fast
        def test_is_palindrome(self, app):
            """Тест проверки палиндромов"""
            # ARRANGE - ПОДГОТОВКА ТЕСТОВОГО ОКРУЖЕНИЯ
            palindrome = "А роза упала на лапу Азора"
            non_palindrome = "hello"

            # ACT - ДЕЙСТВИЕ
            actual_palindrome_result = app.is_palindrome(palindrome)
            actual_non_palindrome_result = app.is_palindrome(non_palindrome)

            # ASSERT
            assert actual_palindrome_result, "Строка 'А роза упала на лапу Азора' должна быть палиндромом"
            assert not actual_non_palindrome_result, "Строка 'hello' не должна быть палиндромом"

        # ГРУППА: STRING_OPERATIONS
        @pytest.mark.string
        @pytest.mark.boundary
        def test_is_palindrome_with_none(self, app):
            """Тест is_palindrome с None"""
            # ARRANGE - ПОДГОТОВКА ТЕСТОВОГО ОКРУЖЕНИЯ
            none_string = None

            # ACT- ДЕЙСТВИЕ
            actual_result = app.is_palindrome(none_string)

            # ASSERT - СРАВНЕНИЕ ОЖИДАЕМОГО И ФАКТИЧЕСКОГО РЕЗУЛЬТАТОВ
            assert not actual_result, "None строка не должна быть палиндромом"

    # Дополнительные boundary тесты
    @pytest.mark.math
    @pytest.mark.boundary
    def test_is_prime_boundary(self, app):
        """Тест boundary случаев для простых чисел"""
        assert not app.is_prime(2147483646), "MAX_VALUE-1 не должно быть простым"

    @pytest.mark.math
    @pytest.mark.boundary
    def test_factorial_boundary(self, app):
        """Тест boundary случаев для факториала"""
        # В Python нет автоматического переполнения для целых чисел,
        # но проверим большие значения
        result = app.factorial(20)
        assert result > 0, "Факториал 20 должен быть положительным числом"

    @pytest.mark.array
    @pytest.mark.boundary
    def test_find_max_with_negative_numbers(self, app):
        """Тест поиска максимального с отрицательными числами"""
        negative_numbers = [-5, -2, -10, -1]
        assert app.find_max(negative_numbers) == -1

File: test/src/test/java/org/example/MainTest.java
package org.example;
import org.testng.annotations.Test;
import static org.testng.Assert.*;

import org.testng.annotations.DataProvider;
//import static org.example.Main.isPrime;
public class MainTest {
    private Main app = new Main();
    @DataProvider(name = "primeNumbers")
    public Object[][] primeNumbersData() {
        return new Object[][] {
                {2, true},
                {3, true},
                {17, true},
                {15, false},
//                {1, false},
//                {0, false},
                {1, true},  // ОШИБКА: 1 не простое число
                {0, true}   // ОШИБКА: 0 не простое число
        };
    }

    @DataProvider(name = "factorialData")
    public Object[][] factorialData() {
        return new Object[][] {
//                {0, 1L},
//                {1, 1L},
//                {5, 120L},
//                {7, 5040L},
                {0, 0L},    // ОШИБКА: факториал 0 должен быть 1
                {1, 3L},    // ОШИБКА: должно быть 2 (1! + 1)
                {5, 125L},  // ОШИБКА: должно быть 121 (120 + 1)
                {7, 5041L}  // ОШИБКА: должно быть 5041 (5040 + 1)
        };
    }

    @DataProvider(name = "palindromeData")
    public Object[][] palindromeData() {
        return new Object[][] {
//                {"А роза упала на лапу Азора", true},
//                {"racecar", true},
//                {"hello", false},
//                {"", true},
//                {"a", true},
                {"А роза упала на лапу Азора", false}, // ОШИБКА: должен быть true
                {"racecar", true},
                {"hello", true},  // ОШИБКА: не палиндром
                {"", false},      // ОШИБКА: пустая строка - палиндром
                {"a", false}      // ОШИБКА: один символ - палиндром
        };
    }

    // ГРУППА: MATH_OPERATIONS
    @Test(groups = {"math", "fast"})
    public void testIsPrime() {
        // ARRANGE - ПОДГОТОВКА ТЕСТОВОГО ОКРУЖЕНИЯ
        int primeNumber = 17;
        int nonPrimeNumber = 15;
        boolean actualPrimeResult;
        boolean actualNonPrimeResult;
        // ACT - ДЕЙСТВИЕ
        actualPrimeResult = app.isPrime(primeNumber);
        actualNonPrimeResult = app.isPrime(nonPrimeNumber);

        // ASSERT - СРАВНЕНИЕ ОЖИДАЕМОГО И ФАКТИЧЕСКОГО РЕЗУЛЬТАТОВ
//        assertTrue(actualPrimeResult, "17 должно быть простым числом");
//        assertFalse(actualNonPrimeResult, "15 не должно быть простым числом");
        // НЕПРАВИЛЬНЫЕ ASSERT
        assertFalse(actualPrimeResult, "17 НЕ должно быть простым числом"); // ОШИБКА
        assertTrue(actualNonPrimeResult, "15 должно быть простым числом");  // ОШИБКА
    }

    // ГРУППА: MATH_OPERATIONS
    @Test(dataProvider = "primeNumbers", groups = {"math"})
    public void testIsPrimeWithDataProvider(int number, boolean expected) {
        // ARRANGE - данные подготовлены в DataProvider

        // ACT - ДЕЙСТВИЕ
        boolean actualResult = app.isPrime(number);

        // ASSERT - СРАВНЕНИЕ ОЖИДАЕМОГО И ФАКТИЧЕСКОГО РЕЗУЛЬТАТОВ
        assertEquals(actualResult, expected,
                "Число " + number + " должно " + (expected ? "быть" : "не быть") + " простым");
    }

    // ГРУППА: MATH_OPERATIONS
    @Test(groups = {"math", "fast"})
    public void testFactorial() {
        // ARRANGE - ПОДГОТОВКА ТЕСТОВОГО ОКРУЖЕНИЯ
        int inputNumber = 5;
//        long expectedResult = 120L;
        long expectedResult = 100L; // ОШИБКА: должно быть 121


        // ACT  - ДЕЙСТВИЕ
        long actualResult = app.factorial(inputNumber);

        // ASSERT - СРАВНЕНИЕ ОЖИДАЕМОГО И ФАКТИЧЕСКОГО РЕЗУЛЬТАТОВ
//        assertEquals(actualResult, expectedResult, "Факториал 5 должен быть равен 120");
        assertEquals(actualResult, expectedResult, "Факториал 5 должен быть равен 100"); // ОШИБКА
    }

    // ГРУППА: MATH_OPERATIONS + EXCEPTION_TESTS
    @Test(expectedExceptions = IllegalArgumentException.class, groups = {"math", "exception"})
    public void testFactorialNegative() {
        // ARRANGE - ПОДГОТОВКА ТЕСТОВОГО ОКРУЖЕНИЯ
        int negativeNumber = -1;

        // ACT - ДЕЙСТВИЕ
        app.factorial(negativeNumber);

        // ASSERT - ожидаем исключение (проверяется в аннотации)
    }
    // ГРУППА: MATH_OPERATIONS
    @Test(dataProvider = "factorialData", groups = {"math"})
    public void testFactorialWithDataProvider(int number, long expected) {
        // ARRANGE - данные подготовлены в DataProvider

        // ACT - ДЕЙСТВИЕ
        long actualResult = app.factorial(number);

        // ASSERT - СРАВНЕНИЕ ОЖИДАЕМОГО И ФАКТИЧЕСКОГО РЕЗУЛЬТАТОВ
        assertEquals(actualResult, expected,
                "Факториал " + number + " должен быть равен " + expected);
    }
    // ГРУППА: ARRAY_OPERATIONS
    @Test(groups = {"array", "fast"})
    public void testFindMax() {
        // ARRANGE - ПОДГОТОВКА ТЕСТОВОГО ОКРУЖЕНИЯ
        int[] numbers = {3, 7, 2, 9, 1};
//        int expectedResult = 9;
        int expectedResult = 7; // ОШИБКА: должно быть 2 (из-за ошибки в методе)
        // ACT - ДЕЙСТВИЕ
        int actualResult = app.findMax(numbers);

        // ASSERT - СРАВНЕНИЕ ОЖИДАЕМОГО И ФАКТИЧЕСКОГО РЕЗУЛЬТАТОВ
//        assertEquals(actualResult, expectedResult, "Максимальное число должно быть 9");
        assertEquals(actualResult, expectedResult, "Максимальное число должно быть 7"); // ОШИБКА
    }

    // ГРУППА: ARRAY_OPERATIONS
    @Test(groups = {"array"})
    public void testFindMaxWithSingleElement() {
        // ARRANGE - ПОДГОТОВКА ТЕСТОВОГО ОКРУЖЕНИЯ
        int[] singleElementArray = {5};
        int expectedResult = 5;

        // ACT - ДЕЙСТВИЕ
        int actualResult = app.findMax(singleElementArray);

        // ASSERT - СРАВНЕНИЕ ОЖИДАЕМОГО И ФАКТИЧЕСКОГО РЕЗУЛЬТАТОВ
        assertEquals(actualResult, expectedResult,
                "Максимальное число в массиве из одного элемента должно быть этим элементом");
    }

    // ГРУППА: ARRAY_OPERATIONS + EXCEPTION_TESTS
    @Test(expectedExceptions = IllegalArgumentException.class, groups = {"array", "exception"})
    public void testFindMaxEmptyArray() {
        // ARRANGE - ПОДГОТОВКА ТЕСТОВОГО ОКРУЖЕНИЯ
        int[] emptyArray = {};

        // ACT - ДЕЙСТВИЕ
        app.findMax(emptyArray);

        // ASSERT - ожидаем исключение (проверяется в аннотации)
    }

    // ГРУППА: STRING_OPERATIONS
    @Test(groups = {"string", "fast"})
    public void testIsPalindrome() {
        // ARRANGE - ПОДГОТОВКА ТЕСТОВОГО ОКРУЖЕНИЯ
        String palindrome = "А роза упала на лапу Азора";
        String nonPalindrome = "hello";

        // ACT - ДЕЙСТВИЕ
        boolean actualPalindromeResult = app.isPalindrome(palindrome);
        boolean actualNonPalindromeResult = app.isPalindrome(nonPalindrome);

        // ASSERT
        assertTrue(actualPalindromeResult, "Строка 'А роза упала на лапу Азора' должна быть палиндромом");
        assertFalse(actualNonPalindromeResult, "Строка 'hello' не должна быть палиндромом");
    }

    // ГРУППА: STRING_OPERATIONS
    @Test(dataProvider = "palindromeData", groups = {"string", "fast"})
    public void testIsPalindromeWithDataProvider(String input, boolean expected) {
        // ARRANGE - данные подготовлены в DataProvider

        // ACT - ДЕЙСТВИЕ
        boolean actualResult = app.isPalindrome(input);

        // ASSERT - СРАВНЕНИЕ ОЖИДАЕМОГО И ФАКТИЧЕСКОГО РЕЗУЛЬТАТОВ
        assertEquals(actualResult, expected,
                "Строка '" + input + "' должна " + (expected ? "быть" : "не быть") + " палиндромом");
    }

    // ГРУППА: STRING_OPERATIONS
    @Test(groups = {"string", "boundary"})
    public void testIsPalindromeWithNull() {
        // ARRANGE - ПОДГОТОВКА ТЕСТОВОГО ОКРУЖЕНИЯ
        String nullString = null;

        // ACT- ДЕЙСТВИЕ
        boolean actualResult = app.isPalindrome(nullString);

        // ASSERT - СРАВНЕНИЕ ОЖИДАЕМОГО И ФАКТИЧЕСКОГО РЕЗУЛЬТАТОВ
        assertFalse(actualResult, "Null строка не должна быть палиндромом");
    }

    @Test(groups = {"math", "fast"})
    public void testFibonacci() {
        // ARRANGE - ПОДГОТОВКА ТЕСТОВОГО ОКРУЖЕНИЯ
        int[] inputs = {0, 1, 6, 10};
        int[] expectedResults = {0, 1, 8, 55};

        // ACT & ASSERT - ДЕЙСТВИЕ И ПРОВЕРКА ДЛЯ НЕСКОЛЬКИХ ЗНАЧЕНИЙ
        for (int i = 0; i < inputs.length; i++) {
            // ACT - ДЕЙСТВИЕ
            int actualResult = app.fibonacci(inputs[i]);

            // ASSERT - СРАВНЕНИЕ ОЖИДАЕМОГО И ФАКТИЧЕСКОГО РЕЗУЛЬТАТОВ
            assertEquals(actualResult, expectedResults[i],
                    "Число Фибоначчи F(" + inputs[i] + ") должно быть равно " + expectedResults[i]);
        }
    }

    // ГРУППА: MATH_OPERATIONS + EXCEPTION_TESTS
    @Test(expectedExceptions = IllegalArgumentException.class, groups = {"math", "exception"})
    public void testFibonacciNegative() {
        // ARRANGE - ПОДГОТОВКА ТЕСТОВОГО ОКРУЖЕНИЯ
        int negativeIndex = -5;

        // ACT- ДЕЙСТВИЕ
        app.fibonacci(negativeIndex);

        // ASSERT - ожидаем исключение (проверяется в аннотации)
    }
}
